towerlib/towerlib.py

```diff
@@ -1997,24 +2034,707 @@ def schedules(self):
+
+        Args:
+            host: the host object.
+
+        Returns:
+            list: list of custom groups.
+
+        """
+        url = "{api}/hosts/{id}/groups/".format(api=self.api, id=host.id)
+        response = self.session.get(url)
+        if not response.ok:
+            self._logger.error("Error getting the project updates. response was: {})".format(response.text))
+            return None
+        else:
+            results = response.json().get('results', [])
+            groups = []
+            for group in results:
+                groups.append(group)
+            return groups
+
+    def get_all_groups_by_host(self, host):
+        """Get groups for a particular host, which are directly and indirectly connected.
+
+        Args:
+            host: the host object.
+
+        Returns:
+            list: list of custom groups.
+
+        """
+        url = "{api}/hosts/{id}/all_groups/".format(api=self.api, id=host.id)
+        response = self.session.get(url)
+        if not response.ok:
+            self._logger.error("Error getting the project updates. response was: {})".format(response.text))
+            return None
+        else:
+            results = response.json().get('results', [])
+            groups = []
+            for group in results:
+                groups.append(group)
+            return groups
+
+    def get_all_groups(self):
+        """Get all the groups of the given ansible tower.
+
+        Returns:
+            list: list of all the groups as EntityManager object.
+
+        """
+        groups = [item for item in self.groups]
+        return groups
+
+
+    def disable_scm_update_on_launch_for_all_projects(self):
+        """Set update_on_launch option for all projects to false.
+
+        """
+        projects = self.get_all_projects()
+        i = 0
+        for project in projects:
+            if project.scm_update_on_launch is True:
+                i += 1
+                project.scm_update_on_launch = False
+        self._logger.debug("{} pojrect(s)'s scm_update_on_lanuch set to False out of total {} projects."
+                           .format(i, len(projects)))
+
+    def search_generic_item_by_keyword(self, generic_item_name_plural, keyword=''):
+        """
+        Search query string parameter to perform a case-insensitive search within all designated text fields of a model
+        or item. Model/item means projects, jobs, inventories etc.
+        https://docs.ansible.com/ansible-tower/latest/html/towerapi/searching.html
+
+        Args:
+            generic_item_name_plural: the plural name of the model/item e.g. projects, jobs, job_templates, users etc.
+            keyword: case insensitive search string.
+
+        Returns:
+            list: list of matching item objects.
+
+        """
+        results = []
+        url = "{api}/{item}/?search={keyword}".format(api=self.api, item=generic_item_name_plural, keyword=keyword)
```

Comment:
A filtering method that validates the argument against a know set of item names with a clear error would be nice here.

Reply:
Something like this a0ceb6990870e632d5355e0acb0429cec40062da ?